File: arrays_and_hashing/typescript/group_anagrams.ts
//https://leetcode.com/problems/group-anagrams/
//
// TODO: find a better solution

const _isAnagram = (s: string, t: string): boolean => {
  if (s.length !== t.length) return false;

  const alphabet = new Uint32Array(26);

  const xs = [...s, ...t];

  for (let i = 0, j = xs.length - 1; i < j; i++, j--) {
    const leftCharCodeIndex = xs[i].charCodeAt(0) - 97
    const rightCharCodeIndex = xs[j].charCodeAt(0) - 97

    alphabet[leftCharCodeIndex]++
    alphabet[rightCharCodeIndex]--
  }
  return alphabet.every(v => v === 0)
}

function groupAnagrams(strs: string[]): string[][] {
  const result: string[][] = []
  while (strs.length) {
    const anagramIndexes: number[] = [];
    const str = strs.shift()
    if (str === undefined) break
    strs.forEach((s, i) => {
      if (_isAnagram(s, str)) {
        anagramIndexes.push(i)
      }
    })

    result.push([str, ...anagramIndexes.map(x => strs[x])])
    strs = strs.filter((_, i) => !anagramIndexes.includes(i))
  }

  return result
};

console.log(groupAnagrams([""]))

File: arrays_and_hashing/go/group-anagrams.go
// https://leetcode.com/problems/group-anagrams/submissions/
// TODO: find a better solution
package main

import (
	"fmt"
)

func main() {
	test := []string{"eat", "tea", "tan", "ate", "nat", "bat"}
	fmt.Println(groupAnagrams(test))
}

func groupAnagrams(strs []string) [][]string {
	checkedIndexes := make(map[int]bool)
	result := [][]string{}

	for i, v := range strs {
		_, ok := checkedIndexes[i]

		if ok {
			continue
		}

		currentGroup := []string{v}
		checkedIndexes[i] = true
		for j := i + 1; j < len(strs); j++ {
			_, ok := checkedIndexes[j]

			if ok {
				continue
			}

			if isAnagram(v, strs[j]) {
				checkedIndexes[j] = true
				currentGroup = append(currentGroup, strs[j])
			}
		}

		result = append(result, currentGroup)
	}

	return result
}

func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	alphabet := [123]int{}

	for i := range s {
		alphabet[int(s[i])]++
	}

	for i := range t {
		alphabet[int(t[i])]--
	}

	for i := 97; i < 123; i++ {
		if alphabet[i] != 0 {
			return false
		}
	}
	return true
}
